src/main/java/com/iexec/sms/tee/session/gramine/GramineSessionMakerService.java

```diff
@@ -0,0 +1,82 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.iexec.sms.tee.session.gramine;
+
+import com.iexec.common.tee.TeeEnclaveProvider;
+import com.iexec.sms.api.config.TeeServicesProperties;
+import com.iexec.sms.tee.ConditionalOnTeeProvider;
+import com.iexec.sms.tee.session.base.SecretEnclaveBase;
+import com.iexec.sms.tee.session.base.SecretSessionBase;
+import com.iexec.sms.tee.session.base.SecretSessionBaseService;
+import com.iexec.sms.tee.session.generic.TeeSessionGenerationException;
+import com.iexec.sms.tee.session.generic.TeeSessionRequest;
+import com.iexec.sms.tee.session.gramine.sps.GramineEnclave;
+import com.iexec.sms.tee.session.gramine.sps.GramineSession;
+import com.iexec.sms.tee.session.gramine.sps.GramineSession.GramineSessionBuilder;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+@ConditionalOnTeeProvider(providers = TeeEnclaveProvider.GRAMINE)
+public class GramineSessionMakerService {
+
+    private final SecretSessionBaseService secretSessionBaseService;
+    private final TeeServicesProperties teeServicesConfig;
+
+    public GramineSessionMakerService(SecretSessionBaseService secretSessionBaseService,
+                                      TeeServicesProperties teeServicesConfig) {
+        this.secretSessionBaseService = secretSessionBaseService;
+        this.teeServicesConfig = teeServicesConfig;
+    }
+
+    /**
+     * Collect tokens required for different compute stages (pre, in, post)
+     * and build the JSON config of the TEE session.
+     *
+     * @param request session request details
+     * @return session config
+     */
+    public GramineSession generateSession(TeeSessionRequest request) throws TeeSessionGenerationException {
+        SecretSessionBase baseSession = secretSessionBaseService.getSecretsTokens(request);
+        GramineSessionBuilder gramineSession = GramineSession.builder()
+                .session(request.getSessionId());
+        GramineEnclave gramineAppEnclave = toGramineEnclave(baseSession.getAppCompute());
+        gramineAppEnclave.setCommand(request.getTaskDescription().getAppCommand());
+        GramineEnclave graminePostEnclave = toGramineEnclave(baseSession.getPostCompute());
+        graminePostEnclave.setCommand(teeServicesConfig.getPostComputeProperties().getEntrypoint());
+
+        // TODO: Remove useless volumes when SPS is ready
+        gramineAppEnclave.setVolumes(List.of());
+        graminePostEnclave.setVolumes(List.of());
```

Comment:
Small reminder, it was discussed previously and it seems to be a better option to use `Collections.emptyList()` instead of `List.of()`.
The rationale was that a single instance is created and reused in Collections class.

Reply:
Fun fact I've just learnt: here is `List.of()` implementation:
```java
    static <E> List<E> of() {
        return ImmutableCollections.emptyList();
    }
```
And `ImmutableCollections.emptyList()` implementation:
```java
    static <E> List<E> emptyList() {
        return (List<E>) ListN.EMPTY_LIST;
    }
```
So, no new object is created :)

Reply:
Nice, I didn't know this. I knew it for `Collections.emptyList()`, but not for `List.of()`.
Thanks